### PR TITLE
Session 5 lir

### PR DESCRIPTION
**Pull request by LirBenShahar**

### CODE REVIEW

File: training/src/app/store/generateJoke.selector.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { createSelector } from "@ngrx/store";
+import { AppState } from "./state";
+
+export const getJokeState     = (state: any) => state.appState as AppState;
+export const getJokeListState = (state: any) => state.appState as AppState;
```
Comment (natanazulay) [inline training/src/app/store/generateJoke.selector.ts]: why the duplication? its the same state (app state)

File: training/src/app/chuck-noris/joke-list/joke-list.component.ts [hunk 2]
```diff
@@ -23,8 +26,13 @@ export class JokeListComponent implements OnInit {
 	public columnsToDisplay: string[]              = ['id', 'category', 'created_at'];
 	public expandedElement: Joke;
 
+	constructor(private store: Store<AppState>) {
+	}
+
 	ngOnInit(): void {
-		this.jokeListTable.data = this.jokeList;
+		this.store.select(selectJokeList).subscribe((jokeList: Joke[]) => {
```
Comment (natanazulay) [inline training/src/app/chuck-noris/joke-list/joke-list.component.ts]: what about unsuscribe?

File: training/src/app/store/app.effect.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+					catchError((err) => of(generateJokeFailed({ payload: err })))
+				)
+			)
+		)
+	);
+
+	constructor(private actions$: Actions, private jokeService: JokeGeneratorService) {
+	}
+
+	private getJokeListFromService(searchedValue: string): Observable<Joke[]> {
```
Comment (natanazulay) [inline training/src/app/store/app.effect.ts]: dont think you need a method for it, if its just 1 line of code - do it directly inside the effect

File: training/src/app/store/app.effect.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		return this.jokeService.getJokeList(searchedValue);
+	}
+
+	private getJokeFromService(): Observable<Joke> {
```
Comment (natanazulay) [inline training/src/app/store/app.effect.ts]: same comment

File: training/src/app/store/app.effect.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+				this.getJokeFromService()
+				.pipe(
+					map((res) => generateJokeSuccess({ payload: res })),
+					catchError((err) => of(generateJokeFailed({ payload: err })))
+				)
+			)
+		)
+	);
+
+	constructor(private actions$: Actions, private jokeService: JokeGeneratorService) {
```
Comment (natanazulay) [inline training/src/app/store/app.effect.ts]: contstuctor should be up and not in the middle of methods

File: training/src/app/store/app.reducer.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	on(getJokeList, (state, action) => {
+		return { ...state, jokeList: action.payload }
+	}),
+	on(getJokeListSuccess, (state, action) => {
+		return {...state, jokeList: action.payload }
+	}),
+	on(getJokeListFailed, (state, action) => {
+		return {...state, jokeListErr: action.payload }
+	}),
+	on(generateJoke, (state) => { return {...state}}),
```
Comment (natanazulay) [inline training/src/app/store/app.reducer.ts]: why do we need it? u dont change anything

File: training/src/app/store/generateJoke.actions.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { createAction, props } from '@ngrx/store';
+import { Joke } from "../chuck-noris/joke";
+
+export const generateJoke        = createAction('generateJoke');
+export const generateJokeSuccess = createAction('generateJoke success', props<{ payload: Joke }>());
+export const generateJokeFailed  = createAction('generateJoke failed', props<{ payload: string }>());
+export const getJokeList         = createAction('getJokeList', props<{ payload: any }>());
```
Comment (natanazulay) [inline training/src/app/store/generateJoke.actions.ts]: why the payload isnt a string?

File: training/src/app/store/app.reducer.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import { createReducer, on } from '@ngrx/store';
+import {
+	generateJoke,
+	generateJokeFailed,
+	generateJokeSuccess,
+	getJokeList, getJokeListFailed,
+	getJokeListSuccess
+} from "./generateJoke.actions";
+export const initialAppState = {
```
Comment (natanazulay) [inline training/src/app/store/app.reducer.ts]: move to state.ts

File: training/src/app/store/app.reducer.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+	joke: null,
```
Comment (natanazulay) [inline training/src/app/store/app.reducer.ts]: undefined instad of null

File: training/src/app/chuck-noris/chuck-noris-main/chuck-norris-main.component.ts [hunk 10]
```diff
@@ -12,31 +16,39 @@ import { ActivatedRoute } from "@angular/router";
 })
 
 export class ChuckNorrisMain implements OnInit {
-	public jokeList$: Observable<Joke[]>;
-	public joke$: Observable<Joke>;
-	public isSearchMode: boolean = false;
+	public jokeList$: Observable<Joke[]> = this.store.select(state => state.jokeList);
```
Comment (natanazulay) [inline training/src/app/chuck-noris/chuck-noris-main/chuck-norris-main.component.ts]: why are you not using the selector u create? selectJoke and selectJokes

File: training/src/app/chuck-noris/chuck-noris-main/chuck-norris-main.component.ts [hunk 11]
```diff
@@ -12,31 +16,39 @@ import { ActivatedRoute } from "@angular/router";
...
+	public joke$: Observable<Joke>       = this.store.select(state => state.joke);
```
Comment (natanazulay) [inline training/src/app/chuck-noris/chuck-noris-main/chuck-norris-main.component.ts]: same comment

File: training/src/app/chuck-noris/chuck-noris-main/chuck-norris-main.component.ts [hunk 12]
```diff
@@ -12,31 +16,39 @@ import { ActivatedRoute } from "@angular/router";
...
 
-	constructor(private activatedRoute: ActivatedRoute, private routingService: RoutingService, private jokeService: JokeGeneratorService) {
+	constructor(private store: Store<AppState>, private activatedRoute: ActivatedRoute,
+				private routingService: RoutingService, private jokeService: JokeGeneratorService) {
 	}
 
 	public ngOnInit(): void {
-		this.isSearchMode = this.activatedRoute.snapshot.data['isSearchMode'];
+		this.isSearchMode = this.activatedRoute.snapshot.data[ 'isSearchMode' ];
+		this.store.select(selectJoke).subscribe((joke: Joke) => {
```
Comment (natanazulay) [inline training/src/app/chuck-noris/chuck-noris-main/chuck-norris-main.component.ts]: not sure what u tried to do there? if you pass it with async pipe in the template, u dont need this

File: training/src/app/chuck-noris/chuck-noris-main/chuck-norris-main.component.ts [hunk 13]
```diff
@@ -12,31 +16,39 @@ import { ActivatedRoute } from "@angular/router";
...
+			this.joke$ = of(joke);
+		})
+
+		this.store.select(selectJokeList).subscribe((jokeList: Joke[]) => {
```
Comment (natanazulay) [inline training/src/app/chuck-noris/chuck-noris-main/chuck-norris-main.component.ts]: same comment

**Review by natanazulay — CHANGES_REQUESTED** [review]

File: training/src/app/chuck-noris/joke-list/joke-list.component.ts [hunk 14]
```diff
@@ -17,14 +21,24 @@ import { animate, state, style, transition, trigger } from '@angular/animations'
...
 	public expandedElement: Joke;
+	public untilDestroy: Subject<void>                             = new Subject<void>();
+
+	constructor(private store: Store<AppState>) {
+	}
 
 	ngOnInit(): void {
-		this.jokeListTable.data = this.jokeList;
+		this.store.select(selectJokeList).subscribe((jokeList: Joke[]) => {
+			this.jokeListTable.data = jokeList;
```
Comment (natanazulay) [inline training/src/app/chuck-noris/joke-list/joke-list.component.ts]: you are already getting this data from @Input() no?

**Review by natanazulay — CHANGES_REQUESTED** [review]

File: training/src/app/chuck-noris/joke-list/joke-list.component.ts (hunk 14 above)
Comment (LirBenShahar) [reply training/src/app/chuck-noris/joke-list/joke-list.component.ts]: deleted the input part the data didn't pass from there properly

File: training/src/app/store/generateJoke.actions.ts (hunk 7 above)
Comment (LirBenShahar) [reply training/src/app/store/generateJoke.actions.ts]: image
here is the error i get when trying to change to string

**Pull request closed without merging**